client/python/gradio_client/templates/discord_chat.py

```diff
@@ -0,0 +1,208 @@
+import asyncio
+import os
+import threading
+from threading import Event
+from typing import Optional
+
+import discord
+import gradio as gr
+from discord import Permissions
+from discord.ext import commands
+from discord.utils import oauth_url
+
+import gradio_client as grc
+from gradio_client.utils import QueueError
+
+event = Event()
+
+DISCORD_TOKEN = os.getenv("DISCORD_TOKEN")
+
+
+async def wait(job):
+    while not job.done():
+        await asyncio.sleep(0.2)
+
+
+def get_client(session: Optional[str] = None) -> grc.Client:
+    client = grc.Client("<<app-src>>", hf_token=os.getenv("HF_TOKEN"))
+    if session:
+        client.session_hash = session
+    return client
+
+
+def truncate_response(response: str) -> str:
+    ending = "...\nTruncating response to 2000 characters due to discord api limits."
+    if len(response) > 2000:
+        return response[: 2000 - len(ending)] + ending
+    else:
+        return response
+
+
+intents = discord.Intents.default()
+intents.message_content = True
+bot = commands.Bot(command_prefix="!", intents=intents)
+
+
+@bot.event
+async def on_ready():
+    print(f"Logged in as {bot.user} (ID: {bot.user.id})")
+    event.set()
+    print("------")
+
+
+thread_to_client = {}
+thread_to_user = {}
+
+
+@bot.command()
+@commands.is_owner()
+async def sync(ctx) -> None:
```

Comment:
Mostly a user experience / QOL suggestion: As brought up in #gradio-technical, you could remove needing to `!sync` / `!chat` with a structure like this:
```
class MyClient(discord.Client):
    """This structure allows slash commands to work instantly (instead of needing to sync global commands for up to an hour)"""

    def __init__(self, *, intents: discord.Intents):
        super().__init__(intents=intents)
        self.tree = app_commands.CommandTree(self)

    async def setup_hook(self):
        # This copies the global commands over to our guild
        self.tree.copy_global_to(guild=MY_GUILD)
        await self.tree.sync(guild=MY_GUILD)  
```
I'm using it here (test ground for HuggingBots): https://huggingface.co/spaces/lunarflu/modbot/blob/main/app.py
And you can see the initial example here: https://github.com/Rapptz/discord.py/blob/master/examples/app_commands/basic.py

Then basically you could fit more cool commands into a smaller space + reduce number of steps needed for each command or test 💪 
![image](https://github.com/gradio-app/gradio/assets/70143200/87a9b659-4ff5-493d-ab02-92b8ee6bf918)

Reply:
Ended up following your suggestion @lunarflu!